File: app/bot/handlers/inline.py
from __future__ import annotations

import logging

from aiogram import Bot, Router
from aiogram.enums import ParseMode
from aiogram.exceptions import TelegramBadRequest, TelegramForbiddenError
from aiogram.types import (
    ChosenInlineResult,
    InlineKeyboardMarkup,
    InlineQuery,
    InlineQueryResultArticle,
    InputMediaAnimation,
    InputMediaPhoto,
    InputMediaVideo,
    InputTextMessageContent,
)

from app.bot.ui import DISABLED_LINK_PREVIEW, original_post_button
from app.providers.base import TweetMedia
from app.services import TweetShareService
from app.utils.urls import extract_first_tweet_url

router = Router(name="inline")
logger = logging.getLogger(__name__)


@router.inline_query()
async def inline_query(query: InlineQuery) -> None:
    parsed = extract_first_tweet_url(query.query or "")
    if parsed is None:
        await query.answer(
            [
                InlineQueryResultArticle(
                    id="invalid-link",
                    title="🔗 Нужна ссылка на пост X/Twitter",
                    description="Например: https://x.com/user/status/123",
                    input_message_content=InputTextMessageContent(
                        message_text="🔗 Пришлите ссылку на пост X/Twitter.",
                        link_preview_options=DISABLED_LINK_PREVIEW,
                    ),
                )
            ],
            cache_time=1,
            is_personal=True,
        )
        return

    await query.answer(
        [
            InlineQueryResultArticle(
                id=f"tweet-{parsed.tweet_id}",
                title="📤 Поделиться постом",
                description=parsed.normalized_url,
                input_message_content=InputTextMessageContent(
                    message_text="⏳ Загрузка поста...",
                    parse_mode=ParseMode.HTML,
                    link_preview_options=DISABLED_LINK_PREVIEW,
                ),
                reply_markup=original_post_button(parsed.normalized_url),
            )
        ],
        cache_time=1,
        is_personal=True,
    )


@router.chosen_inline_result()
async def chosen_inline_result(
    result: ChosenInlineResult,
    bot: Bot,
    tweet_share_service: TweetShareService,
) -> None:
    if not result.inline_message_id:
        return

    parsed = extract_first_tweet_url(result.query or "")
    if parsed is None:
        await _safe_edit(
            bot,
            result.inline_message_id,
            "⚠️ Не удалось распознать ссылку на пост X/Twitter.",
        )
        return

    share = await tweet_share_service.process_url(
        parsed,
        telegram_user_id=result.from_user.id,
        chat_id=None,
        mode="inline",
    )
    if not share.ok or share.post is None:
        await _safe_edit(
            bot,
            result.inline_message_id,
            "⚠️ Не удалось получить пост. Возможно, он удален, приватный или временно недоступен.",
            reply_markup=original_post_button(parsed.normalized_url),
        )
        return

    original_url = share.tweet.url if share.tweet is not None else parsed.normalized_url
    button = original_post_button(original_url)
    if share.post.media:
        media = list(share.post.media)
        caption = share.post.caption_html
        if len(media) >= 2:
            await _try_send_album_private(
                bot, result.from_user.id, media, caption=caption, reply_markup=button
            )
        await _safe_edit_media(
            bot,
            result.inline_message_id,
            media[0],
            caption=caption,
            reply_markup=button,
        )
        return

    await _safe_edit(bot, result.inline_message_id, share.post.html, reply_markup=button)


async def _try_send_album_private(
    bot: Bot,
    user_id: int,
    media: list[TweetMedia],
    *,
    caption: str,
    reply_markup: InlineKeyboardMarkup,
) -> None:
    group = [
        _input_group_media(item, caption if index == 0 else None)
        for index, item in enumerate(media)
    ]
    try:
        await bot.send_media_group(chat_id=user_id, media=group)
    except (TelegramBadRequest, TelegramForbiddenError):
        logger.exception("failed to send album to private chat for user %d", user_id)


def _input_group_media(item: TweetMedia, caption: str | None):
    parse_mode = ParseMode.HTML if caption else None
    if item.type == "photo":
        return InputMediaPhoto(media=item.url, caption=caption, parse_mode=parse_mode)
    if item.type == "video":
        return InputMediaVideo(
            media=item.url,
            caption=caption,
            parse_mode=parse_mode,
            width=item.width,
            height=item.height,
            duration=_duration_seconds(item.duration_ms),
        )
    return InputMediaAnimation(
        media=item.url,
        caption=caption,
        parse_mode=parse_mode,
        width=item.width,
        height=item.height,
        duration=_duration_seconds(item.duration_ms),
    )


async def _safe_edit(
    bot: Bot,
    inline_message_id: str,
    text: str,
    *,
    reply_markup: InlineKeyboardMarkup | None = None,
) -> None:
    try:
        await bot.edit_message_text(
            inline_message_id=inline_message_id,
            text=text,
            parse_mode=ParseMode.HTML,
            link_preview_options=DISABLED_LINK_PREVIEW,
            reply_markup=reply_markup,
        )
    except TelegramBadRequest:
        logger.exception("failed to edit inline message")


async def _safe_edit_media(
    bot: Bot,
    inline_message_id: str,
    item: TweetMedia,
    *,
    caption: str,
    reply_markup: InlineKeyboardMarkup,
) -> None:
    try:
        await bot.edit_message_media(
            inline_message_id=inline_message_id,
            media=_input_media(item, caption),
            reply_markup=reply_markup,
        )
        return
    except TelegramBadRequest:
        logger.exception("failed to edit inline media")
    if item.preview_url:
        try:
            await bot.edit_message_media(
                inline_message_id=inline_message_id,
                media=InputMediaPhoto(
                    media=item.preview_url,
                    caption=caption,
                    parse_mode=ParseMode.HTML,
                ),
                reply_markup=reply_markup,
            )
            return
        except TelegramBadRequest:
            logger.exception("failed to edit inline media preview")
    await _safe_edit(bot, inline_message_id, caption, reply_markup=reply_markup)


def _input_media(item: TweetMedia, caption: str):
    if item.type == "photo":
        return InputMediaPhoto(
            media=item.url,
            caption=caption,
            parse_mode=ParseMode.HTML,
        )
    if item.type == "video":
        return InputMediaVideo(
            media=item.url,
            caption=caption,
            parse_mode=ParseMode.HTML,
            width=item.width,
            height=item.height,
            duration=_duration_seconds(item.duration_ms),
        )
    return InputMediaAnimation(
        media=item.url,
        caption=caption,
        parse_mode=ParseMode.HTML,
        width=item.width,
        height=item.height,
        duration=_duration_seconds(item.duration_ms),
    )


def _duration_seconds(duration_ms: int | None) -> int | None:
    if duration_ms is None:
        return None
    return max(1, round(duration_ms / 1000))

File: src/bot/handlers/inline.ts
import { Composer, GrammyError, type InlineKeyboard } from "grammy";
import type {
  InlineQueryResult,
  InputMediaAnimation,
  InputMediaPhoto,
  InputMediaVideo,
} from "grammy/types";

import type { AppContext } from "@/bot/context";
import {
  DISABLED_LINK_PREVIEW,
  INLINE_THUMBNAIL_INVALID,
  INLINE_THUMBNAIL_SHARE,
  INLINE_THUMBNAIL_SIZE,
  INLINE_THUMBNAIL_THREAD,
  INLINE_THUMBNAIL_TRANSLATE_RU,
  originalPostButton,
} from "@/bot/ui";
import { formatTweet, type TelegramPost } from "@/formatters/telegram";
import { buildRichMessage } from "@/formatters/richMessage";
import type { TweetMedia } from "@/providers/base";
import { languageNameInRussian, translateTweet, TranslationError } from "@/services/translation";
import { extractFirstTweetUrl } from "@/utils/urls";

// Result-id prefixes; order matters when matching since they share a stem.
const THREAD_ID_PREFIX = "tweet-thread-";
const TRANSLATED_ID_PREFIX = "tweet-ru-";
const DEFAULT_ID_PREFIX = "tweet-";

export const inlineComposer = new Composer<AppContext>();

inlineComposer.on("inline_query", async (ctx) => {
  const parsed = extractFirstTweetUrl(ctx.inlineQuery.query ?? "");
  if (parsed === null) {
    await ctx.answerInlineQuery(
      [
        {
          type: "article",
          id: "invalid-link",
          title: "Нужна ссылка на пост X/Twitter",
          description: "Например: https://x.com/user/status/123",
          thumbnail_url: INLINE_THUMBNAIL_INVALID,
          thumbnail_width: INLINE_THUMBNAIL_SIZE,
          thumbnail_height: INLINE_THUMBNAIL_SIZE,
          input_message_content: {
            message_text: "🔗 Пришлите ссылку на пост X/Twitter.",
            link_preview_options: DISABLED_LINK_PREVIEW,
          },
        },
      ],
      { cache_time: 1, is_personal: true },
    );
    return;
  }

  const results: InlineQueryResult[] = [
    {
      type: "article",
      id: `${DEFAULT_ID_PREFIX}${parsed.tweetId}`,
      title: "Поделиться постом",
      description: parsed.normalizedUrl,
      thumbnail_url: INLINE_THUMBNAIL_SHARE,
      thumbnail_width: INLINE_THUMBNAIL_SIZE,
      thumbnail_height: INLINE_THUMBNAIL_SIZE,
      input_message_content: {
        message_text: "⏳ Загрузка поста...",
        parse_mode: "HTML",
        link_preview_options: DISABLED_LINK_PREVIEW,
      },
      reply_markup: originalPostButton(parsed.normalizedUrl),
    },
  ];

  // Offer unrolling the whole thread as a separate result (the user picks).
  if (ctx.settings.threadUnrollEnabled) {
    results.push({
      type: "article",
      id: `${THREAD_ID_PREFIX}${parsed.tweetId}`,
      title: "🧵 Поделиться тредом",
      description: "Развернуть весь тред",
      thumbnail_url: INLINE_THUMBNAIL_THREAD,
      thumbnail_width: INLINE_THUMBNAIL_SIZE,
      thumbnail_height: INLINE_THUMBNAIL_SIZE,
      input_message_content: {
        message_text: "⏳ Загрузка треда...",
        parse_mode: "HTML",
        link_preview_options: DISABLED_LINK_PREVIEW,
      },
      reply_markup: originalPostButton(parsed.normalizedUrl),
    });
  }

  if (ctx.runtimeConfig.russianTranslationEnabled) {
    results.push({
      type: "article",
      id: `${TRANSLATED_ID_PREFIX}${parsed.tweetId}`,
      title: "Отправить на русском (beta)",
      description: parsed.normalizedUrl,
      thumbnail_url: INLINE_THUMBNAIL_TRANSLATE_RU,
      thumbnail_width: INLINE_THUMBNAIL_SIZE,
      thumbnail_height: INLINE_THUMBNAIL_SIZE,
      input_message_content: {
        message_text: "⏳ Переводим пост на русский...",
        parse_mode: "HTML",
        link_preview_options: DISABLED_LINK_PREVIEW,
      },
      reply_markup: originalPostButton(parsed.normalizedUrl),
    });
  }

  await ctx.answerInlineQuery(results, { cache_time: 1, is_personal: true });
});

inlineComposer.on("chosen_inline_result", async (ctx) => {
  const inlineMessageId = ctx.chosenInlineResult.inline_message_id;
  if (!inlineMessageId) return;

  const resultId = ctx.chosenInlineResult.result_id;
  const shareThread = resultId.startsWith(THREAD_ID_PREFIX);
  const translateToRussian =
    !shareThread &&
    resultId.startsWith(TRANSLATED_ID_PREFIX) &&
    ctx.runtimeConfig.russianTranslationEnabled;

  const parsed = extractFirstTweetUrl(ctx.chosenInlineResult.query ?? "");
  if (parsed === null) {
    await safeEditText(ctx, inlineMessageId, "⚠️ Не удалось распознать ссылку на пост X/Twitter.");
    return;
  }

  const share = await ctx.services.tweetShare.processUrl(parsed, {
    telegramUserId: ctx.chosenInlineResult.from.id,
    chatId: null,
    mode: "inline",
    // Only the "thread" result unrolls; the plain/translate results stay single.
    unrollThread: shareThread,
  });
  if (!share.ok || share.post === null) {
    await safeEditText(
      ctx,
      inlineMessageId,
      "⚠️ Не удалось получить пост. Возможно, он удален, приватный или временно недоступен.",
      originalPostButton(parsed.normalizedUrl),
    );
    return;
  }

  const originalUrl = share.tweet?.url ?? parsed.normalizedUrl;
  const button = originalPostButton(originalUrl);

  let post = share.post;
  if (translateToRussian && share.tweet) {
    try {
      const translated = await translateTweet(share.tweet, ctx.translator);
      post = formatTweet(translated.tweet, {
        originalLanguageLabel: languageNameInRussian(translated.sourceLang),
      });
    } catch (error) {
      if (error instanceof TranslationError) {
        console.error("translation failed", error.code, error.message);
      } else {
        console.error("translation unexpected error", error);
      }
      await safeEditText(
        ctx,
        inlineMessageId,
        "⚠️ Не удалось перевести пост. Попробуйте поделиться оригиналом.",
        button,
      );
      return;
    }
  }

  await safeEditRich(ctx, inlineMessageId, post, button);
});

async function safeEditRich(
  ctx: AppContext,
  inlineMessageId: string,
  post: TelegramPost,
  replyMarkup: InlineKeyboard,
): Promise<void> {
  // Edit the placeholder into a Rich Message: long text plus a media carousel.
  try {
    await ctx.api.editMessageTextInline(inlineMessageId, buildRichMessage(post), {
      reply_markup: replyMarkup,
    });
    return;
  } catch (error) {
    if (!(error instanceof GrammyError)) throw error;

    console.error("failed to edit inline rich message", error);
  }

  // Fall back to the legacy single-media / text edit if rich messages are unavailable.
  if (post.media.length > 0) {
    await safeEditMedia(ctx, inlineMessageId, post.media[0]!, post.captionHtml, replyMarkup);
    return;
  }

  await safeEditText(ctx, inlineMessageId, post.html, replyMarkup);
}

async function safeEditText(
  ctx: AppContext,
  inlineMessageId: string,
  text: string,
  replyMarkup?: InlineKeyboard,
): Promise<void> {
  try {
    await ctx.api.editMessageTextInline(inlineMessageId, text, {
      parse_mode: "HTML",
      link_preview_options: DISABLED_LINK_PREVIEW,
      reply_markup: replyMarkup,
    });
  } catch (error) {
    if (!(error instanceof GrammyError)) throw error;

    console.error("failed to edit inline message", error);
  }
}

async function safeEditMedia(
  ctx: AppContext,
  inlineMessageId: string,
  item: TweetMedia,
  caption: string,
  replyMarkup: InlineKeyboard,
): Promise<void> {
  try {
    await ctx.api.editMessageMediaInline(inlineMessageId, inputMedia(item, caption), {
      reply_markup: replyMarkup,
    });
    return;
  } catch (error) {
    if (!(error instanceof GrammyError)) throw error;

    console.error("failed to edit inline media", error);
  }

  if (item.previewUrl) {
    try {
      await ctx.api.editMessageMediaInline(
        inlineMessageId,
        {
          type: "photo",
          media: item.previewUrl,
          caption,
          parse_mode: "HTML",
        },
        { reply_markup: replyMarkup },
      );
      return;
    } catch (error) {
      if (!(error instanceof GrammyError)) throw error;

      console.error("failed to edit inline media preview", error);
    }
  }

  await safeEditText(ctx, inlineMessageId, caption, replyMarkup);
}

function inputMedia(
  item: TweetMedia,
  caption: string,
): InputMediaPhoto | InputMediaVideo | InputMediaAnimation {
  if (item.type === "photo") {
    return { type: "photo", media: item.url, caption, parse_mode: "HTML" };
  }
  if (item.type === "video") {
    return {
      type: "video",
      media: item.url,
      caption,
      parse_mode: "HTML",
      width: item.width ?? undefined,
      height: item.height ?? undefined,
      duration: durationSeconds(item.durationMs) ?? undefined,
    };
  }
  return {
    type: "animation",
    media: item.url,
    caption,
    parse_mode: "HTML",
    width: item.width ?? undefined,
    height: item.height ?? undefined,
    duration: durationSeconds(item.durationMs) ?? undefined,
  };
}

function durationSeconds(durationMs: number | null): number | null {
  if (durationMs === null) return null;
  return Math.max(1, Math.round(durationMs / 1000));
}
